### ISSUE TITLE
DOMPurify Reverses Attributes

### ISSUE THREAD

**Issue by onzag** [cure53/DOMPurify#435]
> Dompurify reverses attributes as they are added

### Background & Context

I am trying to ensure that my content is alphabetically ordered, yet dompurify always ensures to provide my content in the wrong order.

### Bug

Replication steps are simple

```ts
const test = DOMPurify.sanitize('<img class="myimg" src="mysrc"/>');
```

Output is

```
'<img src="mysrc" class="myimg"/>
```

I expect my input to equal my output if no values have changed, I have many hooks in dompurify to ensure values remain in alphabetical order because I do cheap diffing, but the purifier ensures the values get reversed, which is a problem as it makes rich text editors detect a change from what they have created.

**Comment by onzag** [comment]
Real example and how it causes issues

```ts
postProcessingHook(node: HTMLElement) {
  if (node.tagName === "IFRAME") {
    const videoSrc = node.dataset.videoSrc || "";
    const origin = node.dataset.videoOrigin || "";
    cleanAllAttribs(node);

    (node as HTMLIFrameElement).allowFullscreen = true;

     // data-video-origin
      node.dataset.videoOrigin = origin;
      // data-video-src
      node.dataset.videoSrc = videoSrc;

      // frameborder
      (node as HTMLIFrameElement).frameBorder = "0";

    if (origin === "vimeo") {
        node.setAttribute("src", `https://player.vimeo.com/video/${videoSrc}?title=0&byline=0&portrait=0&badge=0`);
      } else if (origin === "youtube") {
        node.setAttribute("src", `https://youtube.com/embed/${videoSrc}?rel=0`);
      }
  }
}
```

This results in values in the exact opposite direction that they have been set.

When my rich text editors sees this and compares the value it attempts to reorder as it doesn't fit its criteria.

**Comment by cure53** [comment]
Heya, that is not DOMPurify but the browser. Some browsers decide to sort attributes differently, which also affects our test suite and forces us to have [lots of redundancy](https://github.com/cure53/DOMPurify/blob/master/test/fixtures/expect.js) :D 

Not sure if we can fix this.

**Comment by onzag** [comment]
@cure53 I don't understand, when i add attributes directly to the node, they get added in the order I have set them, and I'd have expect the postprocessing hook to not reprocess my data.... Is there a way to then ensure alphabetical order?....

**Comment by cure53** [comment]
Nay, I don't think so, sorry. If the attribute order is of high relevance, I think we might not be the right library to use as this is outside our feature-set.

**Comment by bendemboski** [comment]
Just to chime in on this one -- I've tested the latest Chrome and Firefox and they both serialize the attributes in the order they are set, so the fact that `_sanitizeAttributes` processes the attributes [in reverse](https://github.com/cure53/DOMPurify/blob/a16346c86797f81fc498a212facb7a8d9721e849/src/purify.js#L989) and re-adds the valid ones inside the loop guarantees that the attribute order will be reversed on any browser that behave this way. e.g. both of the following assertions will fail

```javascript
assert.equal(
  purify('<a href="#" id="foo"></a>'),
  '<a href="#" id="foo"></a>'
);
assert.equal(
  purify('<a id="foo" href="#"></a>'),
  '<a id="foo" href="#"></a>'
);
```

while both of the following will succeed:

```javascript
assert.equal(
  purify(purify('<a href="#" id="foo"></a>')),
  '<a href="#" id="foo"></a>'
);
assert.equal(
  purify(purify('<a id="foo" href="#"></a>')),
  '<a id="foo" href="#"></a>'
);
```

I understand that guaranteeing a particular attribute order isn't something this library wants to do, but perhaps the fact that it's _guaranteed_ to reverse attributes on the two most popular browsers (and I'd guess others) is worth addressing? Perhaps by building up a list of attributes to restore in the while loop and then restoring them in reverse order once the loop is complete? Setting the attributes in the order in which they originally appeared in `.attributes` rather than the reverse order seems like a pretty reasonable thing to do, no? If you think so, I'd be willing to put the changes in a PR.

**Comment by liam-careerhub** [comment]
It does it when testing in Jest with Node too, it doesn't seem tied to a browser

**Comment by xmedeko** [comment]
See also duplicate (and older) issue #276 and a (refused) PR #277

**Comment by BleddP** [comment]
It would be nice to have the option to preserve attribute order, but I also appreciate this lib is not designed to compare strings. If you wish to check whether any input was sanitized you can use something along these lines which I recently used in a project.

A simple implementation could look something like this:

```
 if (newVal !== sanitized) {
    const removed = DOMPurify.removed.map((item: any) => item.attribute?.name || item.element?.outerHTML)
    if (removed?.length > 0) {
      this.sanitizeError= `Your HTML could be vulnerable through these elements: ${removed.join(', ')}`
    } else {
      this.sanitizeError = ''
    }
  }
```
